projects/plugins/jetpack/modules/masterbar/admin-menu/class-base-admin-menu.php

```diff
@@ -455,6 +455,74 @@ public function override_svg_icons() {
 		}
 	}
 
+	/**
+	 * Hide menus that are unauthorized and don't have visible submenus.
+	 *
+	 * This must be done at the end of menu and submenu manipulation in order to avoid performing this check each time
+	 * the submenus are altered.
+	 */
+	public function hide_unauthorized_menus() {
```

Comment:
TIL that Core does not filter out unauthorized menu items when rendering the URL of the first submenu as menu link ([there is a `current_user_can( $submenu_items[0][1] )` missing here](https://core.trac.wordpress.org/browser/trunk/src/wp-admin/menu-header.php?rev=49193#L152)). That sounds like a bug to me, so it'd be good to maybe create a Trac ticket.

I think this logic runs both when rendering the menu in WP Admin and during requests to the `admin-menu` API endpoint (used for rendering the menu in Calypso). I noted this function slightly duplicates logic from the endpoint. Can we remove it and keep this function as the singe source?

This seems to duplicate some logic we're doing in the `admin-menu` endpoint. In theory, Core already filters out unauthorized menu items in WP Admin, and our `admin-menu` API endpoint already does it as well.

Reply:
It's not actually a bug, imho. The submenus are filtered based on their capabilities. The problem is with the `hide-if-js` css; submenus with that class would not be filtered. Because of this, if you hide the first submenu item the menu would still use that slug as a default URL.

Actually... If you think about it, we could simplify a bit the logic from admin-menu endpoint, I think, by removing those checks and try to use Core's way of filtering those submenus (from what I remember the menu and submenus arrays are not filtered in the admin-menu endpoint). But that's just a theory.

Reply:
> The submenus are filtered based on their capabilities
Except when their links are used in the parent menu. [In that case](https://core.trac.wordpress.org/browser/trunk/src/wp-admin/menu-header.php?rev=49193#L168) Core does not check any capability. 

> This method is used to apply the hide CSS class to the menu

Which in turn will prevent the menu/submenu items from showing up in the menu, so we wouldn't need the logic below, right?

https://github.com/Automattic/jetpack/blob/6d33f193152bcde58d68271908fabacbb571a634/projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-admin-menu.php#L218-L226

Reply:
> Except when their links are used in the parent menu. In that case Core does not check any capability.

hmmm... 🤔  I believe $submenu_items array from menu-header.php contains only submenus with the correct capability for the current user. I thought the same when I've looked over that implementation.

> Which in turn will prevent the menu/submenu items from showing up in the menu, so we wouldn't need the logic below, right?

🤔  The hide logic only applies to our own custom implementation (adding that css class), but this logic is used to check for the WP capabilities and filters out the response. It's unrelated to that implementation 🤔

Reply:
> I believe $submenu_items array from menu-header.php contains only submenus with the correct capability for the current user

You're right! Not sure how I missed that. [The submenus are not added if the user doesn't have the proper capability](https://core.trac.wordpress.org/browser/trunk/src/wp-admin/includes/plugin.php?rev=49193#L1408).

In that case, why do we need the `hide_unauthorized_menus` function?

Reply:
It's needed because we need to simulate this for submenus that are hidden as well. I think an example would be helpful. 
- Let's say you login in wp-admin with a Contributor account. 
- The menu `Tools` contains only the default submenu `tools.php` which is hidden with our CSS class even though the user has the capability to access that page.
- Tools menu contains a link to tools.php since that's the default slug.
- In this case, we should hide the entire menu `Tools` since it would point to tools.php (which has been hidden by our implementation).

Reply:
Ok, that makes sense, but it's not related to any unauthorized submenu, isn't it? The Tools menu is hidden because we don't want to show that WP Admin page to WP.com users, not because the user doesn't have the proper capability. As result, the `tools.php` submenu does not contain any visible submenu and therefore the parent menu is hidden. 

I wrongly assumed that `hide_unauthorized_menus` was handling only unauthorized menus/submenus.

Reply:
Yes, that's correct. I'll rename the method to `hide_parent_of_hidden_submenus`. :)